### DISCUSSION TITLE
Packaging Theos for NixOS

### DISCUSSION THREAD

**Discussion by enricozb** [theos/theos#631] [Show and tell]
I've sort-of managed to get Theos packaged using nix. This is a super hacky solution, but it shows some of what needs to be changed in order for a proper packing with Nix: Specifically:
1. instances of `/bin/bash` or `/usr/bin/perl` need to be changed to the more flexible `/usr/bin/env bash` and `/usr/bin/env perl`.
2. `toolchain/linux/iphone/bin/ld` needs to be modified so it doesn't completely overwrite `LD_LIBRARY_PATH`, as NixOS does not have `/lib` (Not sure exactly what this entails).

This solution is too incomplete to be put in the wiki. I also only use a very small portion of Theos, so I'm sure there are other uses of Theos that this `.nix` file wouldn't support, but probably could with some modifications.

Here is my [theos.nix](https://gist.github.com/enricozb/9dbe928e0d02e45c3aed45bbe98a3eb7).

This `theos.nix` installs the swift toolchain and exclusively the 14.4 iOS SDK, which may not be specifically what you're looking for, so modify it to suit your needs. This `theos.nix` also makes `nic.pl` accessible via `theos-nic` on the `PATH`, but I have not tested it. What I have tested is being able to do something like `THEOS=/nix/store/zq2vffjnmb7zlx0811a11ijv1vzcxp5d-theos/share/theos make package install`, where I build and install a swift iOS app, specifically [this iOS app](https://github.com/enricozb/mu-ios-app) to an iPhone. This is good confirmation that the ad hoc elf patching in the nix script is thorough enough for other Theos projects.

Obviously feel free to reopen this issue if Nix(OS) packaging is of interest.

P.S. Thank you Theos devs for the awesome work.

**Comment by kirb** [comment]
Thanks for posting this! I figured I’d move it from an issue to a discussion for the moment. If we decide to come up with something official/permanent, we can make new issues for that.

**Comment by campbellcole** [comment]
Thank you so much for this derivation. I have updated it to the latest version of Theos as well as changed from SDK 14.4 to 14.5 as 14.4 is no longer in the `sdks` repo.

Also, I came up with a solution for the `$THEOS` problem, which you can see at the bottom of the file. It is set up for zsh, but can be easily adapted for any shell.

There are two versions of this module, one which retains the `theos-nic` script, and one which drops that script in favor of using `$THEOS/bin/nic.pl`.

[With `theos-nic`](https://gist.github.com/campbellcole/b1e99985b4b4282d5fde6cd1af58c092)
[Without `theos-nic`](https://gist.github.com/campbellcole/e66a5bbb8369531c6f75a01d5657150b)

Both work correctly, but since `$THEOS` is properly set, I personally don't see much point in using the module with `theos-nic`.

Thanks again!
